Scrape and Load courses workflow [skip preview]

meganli0818 opened the pull request:
This PR adds the course catalogue parser from the branch TairanJ/parse-catelogue and a new manage.py command named parse_and_load (no arguments needed) to parse courses from the amherst course website and then load them into the database. 

**Some information:**
- change the year/semester you would like to parse the courses from, go into amherst_coursework_backend/amherst_coursework_algo/data/course_catalogue/department_catalogue_links.json and add "/(year)(semester)", i.e. "/2526F" or "/2425S" to the end of each department link.


<!-- Korbit AI PR Description Start -->
## Description by Korbit AI

### What change is being made?
Add a new automated workflow for scraping and loading course data using GitHub Actions, and introduce a Django management command to set up course parsing by modifying department catalog links with the current semester identifier.

### Why are these changes being made?
This change enables the automatic scraping and updating of course data on a set schedule (April 1 and November 1), thus streamlining the process of maintaining up-to-date course information in the database. The introduction of the `course_parser_setup` management command helps efficiently manage the URLs used for parsing, ensuring they are always relevant to the current semester.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`
<!-- Korbit AI PR Description End -->

meganli0818 committed:
Add course parsing, with new command parse_and_load to both parse course data and load it into the database.

meganli0818 committed:
Fixed black formatting

meganli0818 committed:
Update load_courses command path and add beautifulsoup4 to requirements

meganli0818 committed:
Moved parse_course_catalogue to escape testing scope

meganli0818 committed:
Remove unused import statements from load_courses command which was causing tests to fail

DhyeyMavani2003 commented:
> /korbit-review

/korbit-review

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/parse_course_catalogue/parse_course_catalogue.py:
```diff
@@ -0,0 +1,762 @@
+
+def parse_department_catalogue(department_url: str) -> List[str]:
+    """Parse a department page and extract course links."""
+    headers = {
+        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
+        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
+        "Accept-Language": "en-US,en;q=0.5",
+    }
```
@TairanJ @meganli0818 think more here! This is good to explore and hide as a reference to secret key because this is our major (and VITAL) input data stream that needs to be well validated and secure!

meganli0818 committed:
Refactor course parsing commands: remove parse_and_load, add get_all_department_courses, parse_deg_1, and parse_deg_2 commands for improved modularity and clarity.

meganli0818 committed:
Add USER_AGENTS environment variable for secret key purposes, add workflows for parse_courses and load_courses, and removed duplicate parse_course_second_deg function

meganli0818 committed:
Add branch input to workflow dispatch for load_courses and parse_courses

meganli0818 committed:
Switching load_courses and parse_courses worfklows to pull_request trigger (for testing purposes)

meganli0818 committed:
Fixed bug in workflow activation

meganli0818 committed:
Update workflows to use dynamic ref and fetch-depth; enhance user agent handling in request headers

github-actions[bot] committed:
Update database with latest course data [automated]

meganli0818 committed:
Remove pull_request trigger from load_courses and parse_courses workflows

meganli0818 committed:
Add step to fetch latest changes in load_courses and parse_courses workflows

meganli0818 commented:
Sorry guys for all the commits, I couldn't figure out how to test run the workflows locally.

meganli0818 commented:
### NEW ADDITIONS

Added changes Korbit recommended including:
- secret user agent for scraping from Amherst website
- removing duplicate function definitions

Added 2 new workflows:
- load_courses is called twice a year on April 1st 5 AM UTC and November 1st 5 AM UTC
- parse_courses is called twice a year on April 1st 12 AM UTC and November 1st 12 AM UTC (To adjust which semester the courses are parsed for, adjust the SEMESTER global variable in course_parser_setup)

meganli0818 committed:
Add command to generate semester-specific department catalog links

meganli0818 committed:
Add setup command for course parser in parse_courses workflow

meganli0818 commented:
Documentation to come!

meganli0818 committed:
Add documentation for course parser commands and setup

meganli0818 committed:
Fixed black formatting

meganli0818 committed:
Updated black version to 25.1.0 in requirements.txt

meganli0818 commented:
Documentation done!

DhyeyMavani2003 reviewed (approved):
LGTM!
